source/slang/slang-state-serialize.cpp

```diff
@@ -0,0 +1,989 @@
+// slang-state-serialize.cpp
+#include "slang-state-serialize.h"
+
+#include "../core/slang-text-io.h"
+
+#include "../core/slang-math.h"
+
+#include "slang-source-loc.h"
+
+namespace Slang {
+
+
+namespace { // anonymous
+
+struct StoreContext
+{
+    typedef StateSerializeUtil::FileState FileState;
+    typedef StateSerializeUtil::SourceFileState SourceFileState;
+    typedef StateSerializeUtil::PathInfoState PathInfoState;
+
+    StoreContext(RelativeContainer* container)
+    {
+        m_container = container;
+    }
+
+    Safe32Ptr<FileState> findFile(const String& uniqueIdentity)
+    {
+        Safe32Ptr<FileState> file;
+        m_uniqueToFileMap.TryGetValue(uniqueIdentity, file);
+        return file;
+    }
+
+    Safe32Ptr<FileState> addFile(const String& uniqueIdentity, const UnownedStringSlice* content)
```

Comment:
Shouldn't the front-end already have deduped files with the same identity into the same C++ object, such that we could use the `SourceFile` object (or whatever) as the key here?

Reply:
Here one of the ideas for FileState is to allow 'unification' of everything 'file-like'. That could be a SourceFiles - which may or may not be actual files. It could also be things read from the IFileSystem which may or may not be source. Admittidly for the moment I think it can only be source, but it may not be in the future (say a module as IR).